token/core/fabtoken/transfer.go

```diff
@@ -50,108 +50,109 @@ func (s *TransferService) Transfer(ctx context.Context, _ string, _ driver.Owner
-			receivers = append(receivers, output.Owner)
-			outputAuditInfos = append(outputAuditInfos, []byte{})
-			continue
-		}
-		recipients, err := s.Deserializer.Recipients(output.Owner)
-		if err != nil {
-			return nil, nil, errors.Wrap(err, "failed getting recipients")
-		}
-		receivers = append(receivers, recipients...)
-		auditInfo, err := s.Deserializer.GetOwnerAuditInfo(output.Owner, ws)
-		if err != nil {
-			return nil, nil, errors.Wrapf(err, "failed getting audit info for sender identity [%s]", driver.Identity(output.Owner).String())
-		}
-		outputAuditInfos = append(outputAuditInfos, auditInfo...)
+	ws := s.WalletService
+	metadata := &driver.TransferMetadata{
+		Inputs:       nil,
+		Outputs:      nil,
+		ExtraSigners: nil,
 	}
 
-	senderAuditInfos := make([][]byte, 0, len(inputTokens))
-	for _, t := range inputTokens {
+	// inputs
+	for i, t := range inputTokens {
 		auditInfo, err := s.Deserializer.GetOwnerAuditInfo(t.Owner, ws)
 		if err != nil {
 			return nil, nil, errors.Wrapf(err, "failed getting audit info for sender identity [%s]", driver.Identity(t.Owner).String())
 		}
-		senderAuditInfos = append(senderAuditInfos, auditInfo...)
+		metadata.Inputs = append(metadata.Inputs, &driver.TransferInputMetadata{
+			TokenID: tokenIDs[i],
+			Senders: []*driver.AuditableIdentity{
+				{
+					Identity:  t.Owner,
+					AuditInfo: auditInfo,
+				},
+			},
+		})
 	}
 
-	receiverAuditInfos := make([][]byte, 0, len(receivers))
-	for _, receiver := range receivers {
-		if len(receiver) == 0 {
-			receiverAuditInfos = append(receiverAuditInfos, []byte{})
-			continue
+	// outputs
+	outputMetadata := &OutputMetadata{}
+	outputMetadataRaw, err := outputMetadata.Serialize()
+	if err != nil {
+		return nil, nil, errors.Wrapf(err, "failed serializing output information")
+	}
+	for _, output := range outs {
+		var receivers []driver.Identity
+		var receiversAuditInfo [][]byte
+		transferOutputMetadata := &driver.TransferOutputMetadata{
+			OutputMetadata:  outputMetadataRaw,
+			OutputAuditInfo: nil,
+			Receivers:       nil,
 		}
-		auditInfo, err := s.Deserializer.GetOwnerAuditInfo(receiver, ws)
-		if err != nil {
-			return nil, nil, errors.Wrapf(err, "failed getting audit info for recipient identity [%s]", receiver.String())
+
+		if len(output.Owner) == 0 { // redeem
+			receivers = append(receivers, output.Owner)
+			receiversAuditInfo = append(receiversAuditInfo, []byte{})
+		} else {
+			recipients, err := s.Deserializer.Recipients(output.Owner)
+			if err != nil {
+				return nil, nil, errors.Wrap(err, "failed getting recipients")
+			}
+			receivers = append(receivers, recipients...)
+			audiInfo, err := s.Deserializer.GetOwnerAuditInfo(output.Owner, ws)
+			if err != nil {
+				return nil, nil, errors.Wrapf(err, "failed getting audit info for sender identity [%s]", driver.Identity(output.Owner).String())
+			}
+			receiversAuditInfo = append(receiversAuditInfo, audiInfo)
+		}
+		for i, receiver := range receivers {
```

Comment:
this would become
```
receivers, err := s.GetOwnerAuditableIdentities(recipients)
transferOutputMetadata := &driver.TransferOutputMetadata{
 OutputMetadata: outputMetadataRaw,
 Receivers: receivers,
}
```
It's also a bit more clear if we instantiate the struct with all its values instead of modifying it later. The more immutable, the easier to understand.

Reply:
okay